[Merged by Bors] - feat(MeasureTheory/Function): stronger version of Vitali's convergence theorem

igorkhavkine opened the pull request:
Prove a stronger version of Vitali's convergence theorem, which does not require an `IsFiniteMeasure` hypothesis. Instead it uses a new uniform tightness hypothesis. The hypothesis `UnifTight` is implemented closely following the structure of the existing `UnifIntegrable`.

---

- [x] depends on: #14570
- [x] depends on: #14646
- [x] depends on: #14954

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

RemyDegenne commented:
Thanks a lot for the PR! Sorry that it tooks weeks to react to it. The reason it was missed by the reviewers is that it lacks labels. You should add the `awaiting-review` label to signal the PR is ready for other people to look at it. There is a special queue where only PRs with that label appear, and reviewers tend to look at that queue. When a PR is not yet ready for review, please add the label WIP. Basically never leave a PR without a label.
Also, to help sorting the PRs, adding the topic label `t-measure-probability` is helpful.
I have added these labels for you now.

About the PR: it is quite large and contains parts that should be PRed separately: the coercions between measurability properties is a big change in the way we use them, and should have its own PR. Could you open another PR with those changes, and edit the first message of this PR to note that it depends on the other one?

igorkhavkine commented:
Thanks, good to know. I tried following the instructions in the [Contributing Guidelines](https://leanprover-community.github.io/contribute/index.html). I wasn't sure that the labels are something for the submitter to choose.

About splitting out the coercions. The main proof doesn't really use them. So if they are somehow against the Mathlib philosophy, they can just be dropped. Or is the point of the separate PR is that's where such a decision would be made?

RemyDegenne commented:
About the coercions: I don't know yet if they are a good idea or not and a separate PR would be a good place to discuss that.

Making this PR independent of the new coercions would make it simpler to review, so if they are indeed not used much, that would be a good thing to do.

urkud commented:
Could you please use explicit terms in `calc` blocks, so that they're readable without looking at the proofs?

igorkhavkine commented:
> Could you please use explicit terms in `calc` blocks, so that they're readable without looking at the proofs?

Done. I ungolfed the `calc` blocks. Hopefully they are more readable now.

P.S.: Is it good etiquette to merge with the `master` branch when doing updates on a PR? I haven't done it this time.

urkud commented:
I changed the definition so that

- it deals with `ENNReal`s right away, no `Real`s involved;
- it doesn't require `s` to be measurable; existence of a measurable set (and more) follows from this one, see new lemma;

Also, I generalized `.add` lemma so that it doesn't need `1 ≤ p` anymore.
Could you please fix the rest of the file? If you need to get a measurable set from `(hf : UnifTight ...)`, then you call the new `exists_measurableSet_indicator` lemma.

igorkhavkine commented:
Fixed the rest of the file (UnifTight.lean). Changed `exists_measurableSet_indicator` to use ≤ instead of <, to be more consistent with the definition of `UnifTight` (making it easier to fix the rest of the file).

urkud commented:
I asked @RemyDegenne to have another look at the final statement. I'll review the style tomorrow.

igorkhavkine commented:
A note about terminology. Is `UnifTight` an appropriate name for the new hypothesis? I've checked two references [Wikipedia](https://en.wikipedia.org/wiki/Tightness_of_measures) and Fremlin (258Xj, Measure Theory vol.2). Both define the term '(uniformly) tight' for a family of _measures_, with Fremlin not dropping 'uniformly'. For measures, the property refers to the uniform smallness of the measure assigned to the complement of a _compact_ set on a topological space.

The property that I defined didn't have a name in any references on Vitali's theorem that I've seen, so I named it by close analogy with the notion above. At the moment I'm happy with `UnifTight` as being sufficiently descriptive, specific and not contradicting convention (an equally good alternative could be `TightFamily`). If it is worth trying to avoid a future clash with the similarly named but distinct notion for measures, one could use instead `TightFuncFamily` or `<appropriate namespace>.UnifTight`. Alternatively, the notion for measures could be introduced in the future as `MeasureTheory.Measure.UnifTight`. Just some thoughts...

urkud commented:
Let's ask @sgouezel about names and the final statement as well. I will review the style once the main statement is cleared.

RemyDegenne commented:
`UnifTight` seems to be a good name for that notion (`UniformlyTight` could be perhaps better since more explicit and not too long? But `UnifTight` matches our `UnifIntegrable`, so it's good). Such families of functions are called tight in Royden and Fitzpatrick, Real Analysis, 4th ed (page 98). Their Vitali convergence theorem uses it (same page).

The related measure notion could be simply called `Tight`.

RemyDegenne commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+  exact (ENNReal.ofReal_toReal hε_top).symm
+
+
+/-- This lemma is less general than `MeasureTheory.unifIntegrable_finite` which applies to
```
```suggestion
/-- This lemma is less general than `MeasureTheory.unifTight_finite` which applies to
```

RemyDegenne commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+      -- Porting note: Original proof was `simp [← le_antisymm hi' hi]`
+      ext; symm; rw [Fin.coe_ofNat_eq_mod, le_antisymm hi' hi, Nat.mod_succ_eq_iff_lt, Nat.lt_succ]
+
+/-- A finite sequence of Lp functions is uniformly integrable. -/
```
```suggestion
/-- A finite sequence of Lp functions is uniformly tight. -/
```

RemyDegenne commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+
+section VitaliConvergence
+
+/- XXX: In the analogous place in `MeasureTheory.Function.UniformIntegrable`, the measure variable
+   is declared `(μ)` non-implicit. I don't see why, as in all relevant cases it can be
+   deduced from other arguments. -/
+variable {μ : Measure α} {p : ℝ≥0∞}
```
Please make a PR for that change in the UniformIntegrable file!

igorkhavkine replied:
Done! #10599

sgouezel commented on Mathlib/MeasureTheory/Function/UniformIntegrable.lean:
```diff
@@ -143,6 +143,32 @@ protected theorem ae_eq (hf : UnifIntegrable f p μ) (hfg : ∀ n, f n =ᵐ[μ]
+    _ ≤ snorm (s.indicator (f i)) p μ := snorm_indicator_le _
+    _ ≤ ENNReal.ofReal ε := hε _ _ hs hμs
+
+/-- Uniform integrability is preserved by restriction of measure to measurable set. -/
```
```suggestion
/-- Uniform integrability is preserved by restriction of measure to a set. -/
```

sgouezel commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+* `MeasureTheory.UnifTight`:
+  A sequence of functions `f` is uniformly tight if for all `ε > 0`, there
+  exists some measurable set `s` with finite measure such that the Lp-norm of
+  `f i` restricted `sᶜ` is smaller than `ε` for all `i`.
```
```suggestion
  `f i` restricted to `sᶜ` is smaller than `ε` for all `i`.
```

sgouezel commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+## Main definitions
+
+* `MeasureTheory.UnifTight`:
+  A sequence of functions `f` is uniformly tight if for all `ε > 0`, there
```
```suggestion
  A sequence of functions `f` is uniformly tight in `L^p` if for all `ε > 0`, there
```

RemyDegenne commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+  have hεp : 0 < ε ^ p.toReal := by simp only [Real.rpow_pos_of_pos, hε]
+  -- decode Memℒp into a.e. strong measurability and finite snorm
+  obtain ⟨haesmf, hsnf⟩ := hf
+  -- transform snorm to lintegral
+  rw [snorm_eq_snorm' (by assumption) (by assumption)] at hsnf
```
You probably should use [MeasureTheory.snorm_eq_lintegral_rpow_nnnorm](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/LpSeminorm/Basic.html#MeasureTheory.snorm_eq_lintegral_rpow_nnnorm). In general, `snorm'` is not meant to be used outside of the file where it's defined as an intermediate step towards snorm.

igorkhavkine replied:
Done in next update.

sgouezel reviewed (commented):
The naming and the statement look fine to me.

RemyDegenne commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+    {g : α → ℝ≥0∞} (haemg : AEMeasurable g μ) (hg : ∫⁻ a, g a ∂μ < ∞)
+    {ε : ℝ} (hε : 0 < ε) :
+    ∃ (s : Set α) (_ : MeasurableSet s) (_ : μ s < ∞),
+      ∫⁻ a, (sᶜ.indicator g) a ∂μ ≤ ENNReal.ofReal ε := by
```
The usual way of writing such results in mathlib is with set integrals instead of indicators: `∫⁻ a in sᶜ, g a ∂μ ≤ ENNReal.ofReal ε` ([MeasureTheory.lintegral_indicator](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Integral/Lebesgue.html#MeasureTheory.lintegral_indicator) can be used to go from one to the other).

igorkhavkine replied:
Fixed in next update.

RemyDegenne commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+
+/-- A constant function is tight. -/
+theorem unifTight_const {g : α → β} (hp : 1 ≤ p) (hp_ne_top : p ≠ ∞) (hg : Memℒp g p μ) :
+    UnifTight (fun _ : ι => g) p μ := fun ε hε ↦ by
+  by_cases hε_top : ε = ∞
```
Let's enter tactic mode directly. It looks cleaner than the `fun ε hε ↦ by` way you have here:
```suggestion
    UnifTight (fun _ : ι => g) p μ := by
  intro ε hε
  by_cases hε_top : ε = ∞
```

igorkhavkine replied:
I was following Yury's lead with this style. :-) My original code use `intro` everywhere. I myself have no strong opinion here.

RemyDegenne commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+theorem unifTight_subsingleton [Subsingleton ι] (hp_one : 1 ≤ p) (hp_top : p ≠ ∞)
+    {f : ι → α → β} (hf : ∀ i, Memℒp (f i) p μ) : UnifTight f p μ := fun ε hε ↦ by
+  by_cases hε_top : ε = ∞
+  · exact ⟨∅, (by measurability), fun _ => hε_top.symm ▸ le_top⟩
```
```suggestion
  · exact ⟨∅, by measurability, fun _ => hε_top.symm ▸ le_top⟩
```

RemyDegenne commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,514 @@
+  case neg => exact ⟨∅, (by measurability), fun i => False.elim <| hι <| Nonempty.intro i⟩
+  cases' hι with i
+  obtain ⟨s, _, hμs, hfε⟩ := (hf i).snorm_indicator_compl_le hp_one hp_top hrε
+  refine' ⟨s, ne_of_lt hμs, fun j => _⟩
```
Please use `refine` instead of `refine'` (everywhere in the PR).
```suggestion
  refine ⟨s, ne_of_lt hμs, fun j => ?_⟩
```

igorkhavkine replied:
Done in next update.

igorkhavkine commented:
Latest update has responded to all comments.

urkud commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,511 @@
+
+
+/-- Core lemma to be used in `MeasureTheory.Memℒp.snorm_indicator_compl_le`. -/
+theorem lintegral_indicator_compl_le
```
This lemma should go to `MeasureTheory/Integral/Lebesgue`.

igorkhavkine replied:
Moved to `MeasureTheory.Integral.Lebesgue`. Had to introduce some auxiliary lemmas to work with the weaker hypotheses.

igorkhavkine replied:
Here's a short [Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Simplify.20proof.20with.20ENNReal) on how to potentially simplify the new `add_sub_add_comm` lemma in `Data.ENNReal.Real`.

urkud commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,511 @@
+theorem lintegral_indicator_compl_le
+    {g : α → ℝ≥0∞} (haemg : AEMeasurable g μ) (hg : ∫⁻ a, g a ∂μ < ∞)
+    {ε : ℝ} (hε : 0 < ε) :
+    ∃ (s : Set α) (_ : MeasurableSet s) (_ : μ s < ∞),
```
```suggestion
    ∃ s : Set α, MeasurableSet s ∧ μ s < ∞ ∧
```

urkud commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,511 @@
+
+/-- Core lemma to be used in `MeasureTheory.Memℒp.snorm_indicator_compl_le`. -/
+theorem lintegral_indicator_compl_le
+    {g : α → ℝ≥0∞} (haemg : AEMeasurable g μ) (hg : ∫⁻ a, g a ∂μ < ∞)
```
We try to assume `≠ ∞` and deduce `< ∞`. Also, can `MeasureTheory.exists_measurable_le_lintegral_eq` help completely drop measurability assumption?
```suggestion
    {g : α → ℝ≥0∞} (haemg : AEMeasurable g μ) (hg : ∫⁻ a, g a ∂μ ≠ ∞)
```
Another question: is it true for any absolutely continuous measure, not just `μ.withDensity g`?

igorkhavkine replied:
Changed to `≠ ∞`, no problem. I did manage to drop the measurability assumption on `g` as you suggested. But for that I needed to introduce the new lemma `MeasureTheory.exists_measurable_le_set_lintegral_eq_of_integrable` (now in `MeasureTheory.Integral.Lebesgue`), which says that the measurable replacement for `g` is good on every measurable subset, not only globally. But I needed an integrability assumption on `g` for that. It may be possible to get right of the integrability assumption, but I don't see an immediate proof.

igorkhavkine replied:
I'm not sure about `μ.withDensity g` vs any measure absolutely continuous with respect to `µ`. If the measure space is σ-compact, by Radon-Nikodym there is no difference. I don't have a good intuition for what absolute continuity means outside that context.

urkud commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,511 @@
+   to formulate it for `f` valued in `ENNReal`. Could be wrapped with `nnnorm` to make it
+   more general. -/
+theorem tendsto_ENNReal_indicator_lt (f : α → ℝ≥0∞) (x : α) :
+    Tendsto (fun M : ℕ => { x | f x < 1 / (↑M + 1) }.indicator f x) atTop (𝓝 0) := by
```
A generalization: if `M` is a `LinearOrderedCommMonoidWithZero` (in fact, any linear order with `∀ x, 0 ≤ x` but we don't have a typeclass for this), `f : α → M`, and `g : β → M`, then `Tendsto (fun m ↦ {x | f x < m}.indicator f x) (nhds 0) (pure 0)` (and same for `≤`).

igorkhavkine replied:
The generalization may be true, but I'm not sure how that generalization would replace `tendsto_ENNReal_indicator_lt` in the proof of `lintegral_indicator_compl_le`. I've left things as they are for now.

urkud commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,511 @@
+  use s, hms, hμs, hM
+
+/-- A single function that is `Memℒp f p μ` is tight wrt to `μ`. -/
+theorem Memℒp.snorm_indicator_compl_le (hp_one : 1 ≤ p) (hp_top : p ≠ ∞)
```
This lemma should go to one of the files in `Function/LpSeminorm`.

urkud commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,511 @@
+theorem Memℒp.snorm_indicator_compl_le (hp_one : 1 ≤ p) (hp_top : p ≠ ∞)
+    {f : α → β} (hf : Memℒp f p μ)
+    {ε : ℝ} (hε : 0 < ε) :
+    ∃ (s : Set α) (_ : MeasurableSet s) (_ : μ s < ∞),
```
```suggestion
    ∃ s : Set α, MeasurableSet s ∧ μ s < ∞ ∧
```

urkud commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,511 @@
+/-- A single function is tight. -/
+theorem unifTight_subsingleton [Subsingleton ι] (hp_one : 1 ≤ p) (hp_top : p ≠ ∞)
+    {f : ι → α → β} (hf : ∀ i, Memℒp (f i) p μ) : UnifTight f p μ := fun ε hε ↦ by
+  by_cases hε_top : ε = ∞
```
I see that you consider `ε = ∞` separately here and there. Probably, we should change the definition to `∀ ε : ℝ≥0`, not `∀ ε : ℝ≥0∞`, then add some casts to the first 2 lemmas.

igorkhavkine replied:
I don't see any difference between `ε : ℝ≥0` with `0 < ε` and `ε : ℝ` with `0 < ε` (except that ℝ arithmetic is easier to work with), which is what this PR was using originally. The move to `ε : ℝ≥0∞` did make some intermediate inequalities simpler by getting rid of casts. I think the only reason to go back would be some definite `Mathlib` convention for `ε`, `δ` hypotheses, which so far I haven't noticed.

urkud commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,511 @@
+  exact ⟨s, ne_of_lt hμs, fun _ => hgε.trans_eq (ENNReal.ofReal_toReal hε_top)⟩
+
+/-- A single function is tight. -/
+theorem unifTight_subsingleton [Subsingleton ι] (hp_one : 1 ≤ p) (hp_top : p ≠ ∞)
```
The usual trick here is to add `unifTight_of_const` assuming `∀ i j, f i = f j` instead of dealing with `fun _ : ι ↦ g`. Then you prove the `Subsingleton` version by `unifTight_of_const fun i j ↦ congr_arg f (Subsingleton.elim i j)`.

igorkhavkine replied:
Probably this theorem can be gotten rid of altogether. It is only used in one place, `unifTight_fin`. Here I closely replicated the analogous code from `MeasureTheory.Function.UniformIntegrable`. If this code is to be updated, perhaps it would be better to do a common refactor of the `unifTight_*` and `unifIntegrable_*` theorems after this PR is accepted.

igorkhavkine commented:
New update responds to most comments (see individual replies above for details).

urkud commented on Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean:
```diff
@@ -1128,6 +1128,57 @@ theorem _root_.Continuous.memℒp_top_of_hasCompactSupport
+  rw [snorm_eq_lintegral_rpow_nnnorm hp_nz hp_top] at hsnf
+  replace hsnf := (ENNReal.rpow_lt_top_iff_of_pos hirp_pos).mp hsnf
+  -- use core result for lintegral (needs only AEMeasurable), the target estimate will be in `hsfε`
+  obtain ⟨s, hms, hμs, hsfε⟩ := lintegral_indicator_compl_le hsnf.ne hεp
```
```suggestion
  obtain ⟨s, hms, hμs, hsfε⟩ := exists_lintegral_indicator_compl_le hsnf.ne hεp
```

urkud commented:
I golfed some helpers lemmas. I'll go over the main file on Monday.

igorkhavkine commented:
> I golfed some helpers lemmas. I'll go over the main file on Monday.

Should the PR switch over to `< ε` instead of `≤ ε` everywhere? Your last update is mixing the two kinds of inequalities together.

YaelDillies commented on Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:
```diff
@@ -547,6 +547,12 @@ theorem rpow_mul (x : ℝ≥0∞) (y z : ℝ) : x ^ (y * z) = (x ^ y) ^ z := by
       simp [coe_rpow_of_ne_zero h, coe_rpow_of_ne_zero this, NNReal.rpow_mul]
 #align ennreal.rpow_mul ENNReal.rpow_mul
 
+theorem rpow_inv_rpow_self {y : ℝ} (hy : y ≠ 0) (x : ℝ≥0∞) : (x ^ y) ^ (1 / y) = x := by
```
```suggestion
theorem rpow_inv_rpow {y : ℝ} (hy : y ≠ 0) (x : ℝ≥0∞) : (x ^ y) ^ y⁻¹ = x := by
```
to match `Real.rpow_inv_rpow` (yes, I know, `NNReal.rpow_inv_rpow_self` is misnamed and misstated, feel free to fix it if you want). Same below

igorkhavkine replied:
Under NNReal there is
```
@[simp] lemma rpow_rpow_inv {y : ℝ} (hy : y ≠ 0) (x : ℝ≥0) : (x ^ y) ^ y⁻¹ = x := by
@[simp] lemma rpow_inv_rpow {y : ℝ} (hy : y ≠ 0) (x : ℝ≥0) : (x ^ y⁻¹) ^ y = x := by
-- as well as (proved invoking the above lemmas via simp) --
theorem rpow_inv_rpow_self {y : ℝ} (hy : y ≠ 0) (x : ℝ≥0) : (x ^ y) ^ (1 / y) = x := by
theorem rpow_self_rpow_inv {y : ℝ} (hy : y ≠ 0) (x : ℝ≥0) : (x ^ (1 / y)) ^ y = x := by
```
The only difference in the names is I guess justified by the difference is in the notation for the inverse. Just to confirm, your recommendation is to just keep the first pair, rather than extend the symmetry to Real and ENNReal?

YaelDillies replied:
Yes, keep only the first pair

YaelDillies commented on Mathlib/Data/ENNReal/Operations.lean:
```diff
@@ -473,6 +473,12 @@ theorem sub_le_sub_iff_left (h : c ≤ a) (h' : a ≠ ∞) :
     (a - b ≤ a - c) ↔ c ≤ b :=
   (cancel_of_ne h').tsub_le_tsub_iff_left (cancel_of_ne (ne_top_of_le_ne_top h' h)) h
 
+protected theorem add_sub_add_comm {a b c d : ℝ≥0∞} (hc : c ≠ ∞) (hd : d ≠ ∞)
+    (hac : c ≤ a) (hbd : d ≤ b) : (a + b) - (c + d) = (a - c) + (b - d) := by
+  refine add_tsub_add_le_tsub_add_tsub.antisymm ?_
+  apply le_sub_of_add_le_left (add_ne_top.2 ⟨hc, hd⟩)
+  rw [add_add_add_comm, add_tsub_cancel_of_le hac, add_tsub_cancel_of_le hbd]
```
Can you instead deduce this from a new lemma `AddLECancellable.add_tsub_add_comm`?

YaelDillies reviewed (commented):
You would get better feedback if you split off the prerequisites. I see at least three separate prerequisite PRs you could open:
* The new lemmas in `Analysis.SpecialFunctions.Pow.NNReal`
* The new lemma in `Data.ENNReal.Operations`
* The various measure theoretic prerequisites.

You can mark this PR as depending on them by writing `- [ ] depends on: #PRNumber` in the PR description.

igorkhavkine commented:
> You would get better feedback if you split off the prerequisites. I see at least three separate prerequisite PRs you could open:
> 
>     * The new lemmas in `Analysis.SpecialFunctions.Pow.NNReal`
>     * The new lemma in `Data.ENNReal.Operations`
>     * The various measure theoretic prerequisites.

It turns out that the new lemmas that I had added in `Data.ENNReal.Operations` and in `Analysis.SpecialFunctions.Pow.NNReal` became unnecessary after the refactoring that was already done earlier. So I just removed them. I think that the only remaining changes that are outside the new `UnifTight.lean` file were moved there through the refactoring and review carried out by @urkud.

I can of course still split the PR, but perhaps Yury can chime in on whether that's really necessary.

YaelDillies commented on Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:
```diff
@@ -707,6 +707,9 @@ theorem le_rpow_one_div_iff {x y : ℝ≥0∞} {z : ℝ} (hz : 0 < z) : x ≤ y
   rw [rpow_mul, ← one_div, @rpow_le_rpow_iff _ _ (1 / z) (by simp [hz])]
 #align ennreal.le_rpow_one_div_iff ENNReal.le_rpow_one_div_iff
 
+theorem rpow_one_div_lt_iff {x y : ℝ≥0∞} {z : ℝ} (hz : 0 < z) : x ^ (1 / z) < y ↔ x < y ^ z := by
```
Please never use `1 / x`. This spelling is deprecated (although not `deprecated`) and we would like no new lemma to use it.
```suggestion
theorem rpow_inv_lt_iff {x y : ℝ≥0∞} {z : ℝ} (hz : 0 < z) : x ^ z⁻¹ < y ↔ x < y ^ z := by
```
Also, when you add such a lemma, please make sure you add all variants, eg here `y < x ^ z⁻¹`, `y ≤ x ^ z⁻¹`, `x ^ z⁻¹ ≤ y`. (I know the variants of the lemma you want to add already exist, this is just general advice).

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,372 @@
+  rw [sub_eq_add_neg]
+  exact hf.add hg.neg hf_meas fun i => (hg_meas i).neg
+
+
```
```suggestion
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,372 @@
+  exact ⟨s, ne_of_lt hμs, fun _ => hgε.le⟩
+
+/-- A single function is tight. -/
+theorem unifTight_subsingleton [Subsingleton ι] (hp_top : p ≠ ∞)
```
```suggestion
theorem unifTight_of_subsingleton [Subsingleton ι] (hp_top : p ≠ ∞)
```

igorkhavkine replied:
The parallel naming convention could also be fixed in UniformIntegrable.lean.

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,372 @@
+  convert hfε.le
+
+
+/-- This lemma is less general than `MeasureTheory.unifTight_finite` which applies to
+all sequences indexed by a finite type. -/
+theorem unifTight_fin (hp_top : p ≠ ∞) {n : ℕ} {f : Fin n → α → β}
```
Maybe make it `private` since nobody (but you when proving `unifTight_finite`) should ever need to use it?

igorkhavkine replied:
OK. A parallel change could be made also in UniformIntegrable.lean.

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,372 @@
+  obtain ⟨s, _, hμs, hfε⟩ := (hf i).exists_snorm_indicator_compl_lt hp_top hε.ne'
+  refine ⟨s, ne_of_lt hμs, fun j => ?_⟩
+  convert hfε.le
+
```
```suggestion
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,372 @@
+
+variable {f : ℕ → α → β} {g : α → β}
+
+/- Both directions and an iff version of Vitali's convergence theorem on measure spaces
+   of not necesserily finite volume. See `Thm III.6.15` of Dunford & Schwartz, Part I (1958). -/
```
If you want it to show up in the docs:
```suggestion
/-! Both directions and an iff version of Vitali's convergence theorem on measure spaces
   of not necesserily finite volume. See `Thm III.6.15` of Dunford & Schwartz, Part I (1958). -/
```
Also please don't indent docstrings/module docs/comments by default

igorkhavkine replied:
Thanks, good to know!

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,372 @@
+    _ ≤ (ε / 3 + ε / 3) + ε / 3 := add_le_add hfngEcε hfngEε
+    _ = ε := by simp only [ENNReal.add_thirds] --ENNReal.add_thirds ε
+
+/- Lemma used in `tendsto_Lp_notFinite_of_tendsto_ae`.
+   XXX: Alternative name: `ae_tendsto_ae_congr`? -/
+theorem tendsto_ae_congr_ae {f f' : ℕ → α → β} {g g' : α → β}
+    (hff' : ∀ (n : ℕ), f n =ᵐ[μ] f' n) (hgg' : g =ᵐ[μ] g')
+    (hfg : ∀ᵐ x ∂μ, Tendsto (fun n => f n x) atTop (𝓝 (g x))) :
+    ∀ᵐ x ∂μ, Tendsto (fun n => f' n x) atTop (𝓝 (g' x)) := by
```
This has nothing to do with `UnifTight`, right? Can you move it to an earlier file? unless it's supposed to be an auxiliary lemma? In that case, please make it `private`.

igorkhavkine replied:
OK, not sure where else it would fit. I will make it `private`.

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,372 @@
+      hms'⟩
+
+
+/-- **Vitali's convergence theorem** (non-finite measure version):
+    A sequence of functions `f` converges to `g` in Lp
+    if and only if it is uniformly integrable, uniformly tight and to `g` in measure. -/
```
```suggestion
/-- **Vitali's convergence theorem** (non-finite measure version).

A sequence of functions `f` converges to `g` in Lp
if and only if it is uniformly integrable, uniformly tight and to `g` in measure. -/
```
to follow the short docstring-long docstring style.

igorkhavkine replied:
OK. A similar update could be done in UniformIntegrable.lean.

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,372 @@
+
+
+end VitaliConvergence
+
+
```
```suggestion
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UniformIntegrable.lean:
```diff
@@ -143,6 +143,32 @@ protected theorem ae_eq (hf : UnifIntegrable f p μ) (hfg : ∀ n, f n =ᵐ[μ]
   simp_rw [Set.indicator_apply, hx]
 #align measure_theory.unif_integrable.ae_eq MeasureTheory.UnifIntegrable.ae_eq
 
+/-- Uniform integrability is preserved by restriction of functions to a set. -/
```
```suggestion
/-- Uniform integrability is preserved by restriction of the functions to a set. -/
```
is clearer maybe?

YaelDillies commented on Mathlib/MeasureTheory/Function/UniformIntegrable.lean:
```diff
@@ -143,6 +143,32 @@ protected theorem ae_eq (hf : UnifIntegrable f p μ) (hfg : ∀ n, f n =ᵐ[μ]
+    _ ≤ snorm (s.indicator (f i)) p μ := snorm_indicator_le _
+    _ ≤ ENNReal.ofReal ε := hε _ _ hs hμs
+
+/-- Uniform integrability is preserved by restriction of measure to a set. -/
```
And similarly
```suggestion
/-- Uniform integrability is preserved by restriction of the measure to a set. -/
```

YaelDillies commented on Mathlib/MeasureTheory/Measure/MeasureSpace.lean:
```diff
@@ -1897,6 +1897,13 @@ theorem eventually_cofinite {p : α → Prop} : (∀ᶠ x in μ.cofinite, p x) 
   Iff.rfl
 #align measure_theory.measure.eventually_cofinite MeasureTheory.Measure.eventually_cofinite
 
+instance : IsMeasurablyGenerated μ.cofinite where
```
```suggestion
instance cofinite.instIsMeasurablyGenerated : IsMeasurablyGenerated μ.cofinite where
```

YaelDillies commented:
> It turns out that the new lemmas that I had added in `Data.ENNReal.Operations` and in `Analysis.SpecialFunctions.Pow.NNReal` became unnecessary after the refactoring that was already done earlier. So I just removed them.

Feel free to open independent PRs for them if you think they are generally useful, though (it happens to me a lot that I write lemmas that I end up not needing, and I usually do PR them anyway).

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,369 @@
+   are unwrapped and strengthened to by known lemmas to have in addition `StronglyMeasurable`
+   and a.e. convergence. The bulk of the proof is done under these stronger hyptheses. -/
+
+theorem tendsto_Lp_notFinite_of_tendsto_ae_of_meas (hp : 1 ≤ p) (hp' : p ≠ ∞)
```
Uh, 
```suggestion
theorem tendsto_Lp_infinite_of_tendsto_ae_of_meas (hp : 1 ≤ p) (hp' : p ≠ ∞)
```
?
But there is no “infinite” in the statement so maybe
```suggestion
theorem tendsto_Lp_of_tendsto_ae_of_meas (hp : 1 ≤ p) (hp' : p ≠ ∞)
```

YaelDillies replied:
Could you make the change in a *preliminary PR*? It will be strictly better in the history to see
* rename `B` to `C`
* introduce `B`

rather than
* introduce `A`
* rename `B` to `C` and `A` to `B`

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,369 @@
+    _ = ε := by simp only [ENNReal.add_thirds] --ENNReal.add_thirds ε
+
+/- Lemma used in `tendsto_Lp_notFinite_of_tendsto_ae`.
+   XXX: Alternative name: `ae_tendsto_ae_congr`? -/
```
I prefer this alternative, yes. But doesn’t hugely matters since it’s private

YaelDillies commented on Mathlib/MeasureTheory/Integral/Lebesgue.lean:
```diff
@@ -1432,6 +1432,11 @@ theorem lintegral_add_compl (f : α → ℝ≥0∞) {A : Set α} (hA : Measurabl
   rw [← lintegral_add_measure, Measure.restrict_add_restrict_compl hA]
 #align measure_theory.lintegral_add_compl MeasureTheory.lintegral_add_compl
 
+theorem set_lintegral_compl {f : α → ℝ≥0∞} {s : Set α} (hsm : MeasurableSet s)
```
```suggestion
theorem setLintegral_compl {f : α → ℝ≥0∞} {s : Set α} (hsm : MeasurableSet s)
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/Lebesgue.lean:
```diff
@@ -1580,6 +1585,51 @@ theorem setLIntegral_subtype {s : Set α} (hs : MeasurableSet s) (t : Set s) (f
+such that the integral of `f` over `sᶜ` is less than a given positive number.
+
+Also used to prove an `Lᵖ`-norm version in `MeasureTheory.Memℒp.exists_snorm_indicator_compl_le`. -/
+theorem exists_set_lintegral_compl_lt {f : α → ℝ≥0∞} (hf : ∫⁻ a, f a ∂μ ≠ ∞)
```
Same here

igorkhavkine commented on Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:
```diff
@@ -734,6 +734,9 @@ theorem le_rpow_one_div_iff {x y : ℝ≥0∞} {z : ℝ} (hz : 0 < z) : x ≤ y
   rw [rpow_mul, ← one_div, @rpow_le_rpow_iff _ _ (1 / z) (by simp [hz])]
 #align ennreal.le_rpow_one_div_iff ENNReal.le_rpow_one_div_iff
 
+theorem rpow_inv_lt_iff {x y : ℝ≥0∞} {z : ℝ} (hz : 0 < z) : x ^ z⁻¹ < y ↔ x < y ^ z := by
```
It is now #14570. Not sure if any further action is required.

YaelDillies commented:
Can you also split off the definition of uniform tightness to a preliminary PR?

igorkhavkine commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,369 @@
+/-! Both directions and an iff version of Vitali's convergence theorem on measure spaces
+   of not necesserily finite volume. See `Thm III.6.15` of Dunford & Schwartz, Part I (1958). -/
+
+/- We start with the reverse direction. We only need to show that uniform tightness follows
```
Hmm, I see this more of an internal comment about the implementation, rather than a public facing doc-string. Same for the later "/- Next we deal with ... -/" comment.

YaelDillies replied:
Sure. The point about not indenting still stands, though

igorkhavkine commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,369 @@
+section UnifTight
+
+/- This follows closely the `UnifIntegrable` section
+   from `MeasureTheory.Functions.UniformIntegrable`.-/
```
Maybe I'm confused about what doc-comments are for, but this one is just a comment about the internal implementation and is not meant to be user-facing.

YaelDillies replied:
Okay sure

urkud commented:
I'm sorry for a long silence. I unassigned this PR from me, because I've failed to review it in time.

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,191 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
 
 end UnifTight
 
+
```
```suggestion
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,191 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
+
+section VitaliConvergence
+
+variable {μ : Measure α} {p : ℝ≥0∞}
+
+variable {f : ℕ → α → β} {g : α → β}
```
```suggestion
variable {μ : Measure α} {p : ℝ≥0∞} {f : ℕ → α → β} {g : α → β}
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,191 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
+      unifIntegrable_of_tendsto_Lp hp hp' hf hg h,
+      unifTight_of_tendsto_Lp hp' hf hg h⟩⟩
+
+
```
```suggestion
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -0,0 +1,369 @@
+
+
+/- Next we deal with the forward direction. The `Memℒp` and `TendstoInMeasure` hypotheses
+   are unwrapped and strengthened to by known lemmas to have in addition `StronglyMeasurable`
```
Can you please resolve conversations when they are not relevant anymore?

igorkhavkine commented:
All PR dependencies have been accepted and merged. the finite/nonFinite renaming has been completed. A couple of lemmas from the UnifTight API strengthened (according to the easy hypotheses/hard theorems philosophy).

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -70,15 +78,16 @@ theorem unifTight_iff_real {_ : MeasurableSpace α} (f : ι → α → β) (p :
     ∀ᶠ s in μ.cofinite.smallSets, ∀ i, snorm (s.indicator (f i)) p μ ≤ ε := by
-  rcases hf (pos_iff_ne_zero.2 hε) with ⟨s, hμs, hfs⟩
-  refine (eventually_smallSets' ?_).2 ⟨sᶜ, ?_, fun i ↦ hfs i⟩
+  by_cases hε_top : ε = ∞; filter_upwards; exact fun _ _ ↦ hε_top.symm ▸ le_top
```
```suggestion
  by_cases hε_top : ε = ∞; subst hε_top; simp
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,185 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
+    _ ≤ (ε / 3 + ε / 3) + ε / 3 := add_le_add hfngEcε hfngEε
+    _ = ε := by simp only [ENNReal.add_thirds] --ENNReal.add_thirds ε
+
+/- Lemma used in `tendsto_Lp_of_tendsto_ae`. -/
```
Might as well make it a docstring
```suggestion
/-- Lemma used in `tendsto_Lp_of_tendsto_ae`. -/
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,185 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
+  apply Filter.Tendsto.congr (fun n => (hsnfg n).symm)
+  exact tendsto_Lp_of_tendsto_ae_of_meas hp hp' hf hg hg'' hui' hut' haefg'
+
+
```
```suggestion
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,185 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
+        let ⟨s, hμs, hfε⟩ := hut hε
+        ⟨s, hμs, fun i => hfε _⟩)
+      hms'⟩
+
```
```suggestion
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,185 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
+theorem tendstoInMeasure_iff_tendsto_Lp (hp : 1 ≤ p) (hp' : p ≠ ∞)
+    (hf : ∀ n, Memℒp (f n) p μ) (hg : Memℒp g p μ) :
+    TendstoInMeasure μ f atTop g ∧ UnifIntegrable f p μ ∧ UnifTight f p μ
+      ↔ Tendsto (fun n => snorm (f n - g) p μ) atTop (𝓝 0) :=
+  ⟨fun h => tendsto_Lp_of_tendstoInMeasure hp hp' (fun n => (hf n).1) hg h.2.1 h.2.2 h.1,
+    fun h =>
+    ⟨tendstoInMeasure_of_tendsto_snorm (lt_of_lt_of_le zero_lt_one hp).ne'
+        (fun n => (hf n).aestronglyMeasurable) hg.aestronglyMeasurable h,
+      unifIntegrable_of_tendsto_Lp hp hp' hf hg h,
+      unifTight_of_tendsto_Lp hp' hf hg h⟩⟩
```
`Iff` is a structure, so you get to do
```suggestion
      ↔ Tendsto (fun n => snorm (f n - g) p μ) atTop (𝓝 0) where
  mp h := tendsto_Lp_of_tendstoInMeasure hp hp' (fun n => (hf n).1) hg h.2.1 h.2.2 h.1
  mpr h := ⟨tendstoInMeasure_of_tendsto_snorm (lt_of_lt_of_le zero_lt_one hp).ne'
        (fun n => (hf n).aestronglyMeasurable) hg.aestronglyMeasurable h,
      unifIntegrable_of_tendsto_Lp hp hp' hf hg h,
      unifTight_of_tendsto_Lp hp' hf hg h⟩
```
which I find much nicer

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -23,6 +26,11 @@ that works also for spaces of infinite measure.
 * `MeasureTheory.unifTight_finite`: a finite sequence of Lp functions is uniformly
   tight.
+* `MeasureTheory.tendsto_Lp_of_tendsto_ae`: a sequence of Lp functions which is uniformly
+  integrable and uniformly tight converges in Lp if it converge almost everywhere.
```
```suggestion
  integrable and uniformly tight converges in Lp if it converges almost everywhere.
```

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,182 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
+are unwrapped and strengthened (by known lemmas) to also have the `StronglyMeasurable`
+and a.e. convergence hypotheses. The bulk of the proof is done under these stronger hypotheses.-/
+
+theorem tendsto_Lp_of_tendsto_ae_of_meas (hp : 1 ≤ p) (hp' : p ≠ ∞)
```
This one is completely superseded by the next, right? If so, can you mark it private?

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,182 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
+and `tendstoInMeasure_of_tendsto_snorm`. -/
+
+/-- Intermediate lemma for `unifTight_of_tendsto_Lp`. -/
+theorem unifTight_of_tendsto_Lp_zero (hp' : p ≠ ∞) (hf : ∀ n, Memℒp (f n) p μ)
```
This one too can be private right?

YaelDillies commented on Mathlib/MeasureTheory/Function/UnifTight.lean:
```diff
@@ -199,4 +207,184 @@ theorem unifTight_finite [Finite ι] (hp_top : p ≠ ∞) {f : ι → α → β}
+  · exact unifTight_const hp' hg
+  · exact unifTight_of_tendsto_Lp_zero hp' (fun n => (hf n).sub hg) hfg
+
+
```
```suggestion
```

YaelDillies reviewed (commented):
Looks good but I would like a measure theorist to sign off.

maintainer merge

semorrison commented:
bors merge
